target/linux/generic/files/drivers/net/phy/mvsw61xx.c

```diff
@@ -670,16 +670,30 @@ static int _mvsw61xx_reset(struct switch_dev *dev, bool full)
 					    BMCR_SPEED1000);
 		}
 
-		/* enable SerDes if necessary */
-		if (full && i >= 5 && state->model == MV_IDENT_VALUE_6176) {
+		if (state->model == MV_IDENT_VALUE_6176) {
 			u16 sts = sr16(dev, MV_PORTREG(STATUS, i));
 			u16 mode = sts & MV_PORT_STATUS_CMODE_MASK;
 
-			if (mode == MV_PORT_STATUS_CMODE_100BASE_X ||
-			    mode == MV_PORT_STATUS_CMODE_1000BASE_X ||
-			    mode == MV_PORT_STATUS_CMODE_SGMII) {
+			/* enable SerDes if necessary */
+			if (i >= 5 && full &&
+				(mode == MV_PORT_STATUS_CMODE_100BASE_X ||
+				 mode == MV_PORT_STATUS_CMODE_1000BASE_X ||
+				 mode == MV_PORT_STATUS_CMODE_SGMII)) {
 				mvsw61xx_enable_serdes(dev);
 			}
+
+			/* Enable RGMII internal delay for CPU ports */
+			if (i == state->cpu_port0 || i == state->cpu_port1) {
```

Comment:
Are you sure this will not harm other devices?
I think the delay has to be activated depending on the PCB design. If the wires are too short the mac can add an additional delay. I have heard of multiple other  non mvebu devices not working correctly when such a delay is accidentally activated or deactivated.

Reply:
No idea about this other than what it says in the commit log...

Reply:
Honestly, I am not sure. My understanding is that there is a minimum delay, and that not all RGMII endpoints contain a delay line as this incurs additional cost (negligible in 2018). I am not aware of a maximum delay.

I assume that the other cases you have heard about are, like the Omnia, not working when the delay is deactivated and again working when it is reactivated. Again, I am not sure.

If there is some documented precedent of devices with RGMII that stops working when delay is *added*, then I think the easiest alternative would be to configure the internal delay on the SoC side and then diff out the DSA switch definition from the device tree as I belive this conflicts with manually configuring link parameters for eth0/eth1.

Reply:
I'm really not sure if it is applicable here, but there are phy-mode device tree values, which are meant to be used if a RGMII delay is required:

* "rgmii" (RX and TX delays are added by the MAC when required)
 * "rgmii-id" (RGMII with internal RX and TX delays provided by the PHY, the  MAC should not add the RX or TX delays in this case)
* "rgmii-rxid" (RGMII with internal RX delay provided by the PHY, the MAC should not add an RX delay in this case)
* "rgmii-txid" (RGMII with internal TX delay provided by the PHY, the MAC should not add an TX delay in this case)

Support for the properties need to be implemented in the switch driver of course. This way it would be possible to enable the delay for boards where required.

Reply:
It would be better to activate this with a device tree property, but I am also fine with waiting till someone complains. ;-)

Reply:
I talked with @mkresin about this topic and we see two solutions here.
1. make the delay activation configurable, e.g. by introducing "cpu-phy-mode-0 = "rgmii-id";
2. Wait till mvebu is migrated to kernel 4.14 and use the DSA switch driver. It already supports this delay, see https://github.com/torvalds/linux/commit/a0a0f6229b81c6398b6614767a2e227c9224a38b this has the disadvantage that only one CPU port will be supported and not two. We want to migrate to DSA in the end anyway.  The current 4.14 patches for mvebu can be found here: https://patchwork.ozlabs.org/project/lede/list/?series=24581